src/compiler/checker.ts

```diff
@@ -4853,6 +4853,13 @@ module ts {
                                 narrowedType = narrowType(type, (<ConditionalExpression>node).condition, /*assumeTrue*/ child === (<ConditionalExpression>node).whenTrue);
                             }
                             break;
+                        case SyntaxKind.CaseClause:
+                            // In a case clause of a switch statement using typeof 
+                            var parent = (<CaseClause>node).parent;
+                            if((<SwitchStatement>parent).expression.kind === SyntaxKind.TypeOfExpression){
+                                narrowedType = checkExpression((<TypeOfExpression>(<CaseClause>node).expression).text);

```

Comment:
This is not quite correct. You essentially need to take narrowTypeByEquality, and factor out the part that that deals with typeof, so you can access that part here. You need to check that the node.expression is a StringLiteral, and then cast it to a LiteralExpression to get the text. The TypeOfExpression in this case would be parent.expression.